File: sdk/python/reconsync/webhook.py
"""Verifying a ReconSync webhook, which is the one thing every receiver must
get right.

The payload advises reversing a customer's money. A handler that acts on an
unverified body will act on anything anyone posts to that URL, and the whole
design — advisory payloads, no credentials in the webhook — rests on the
receiver checking who sent it.
"""

from __future__ import annotations

import hmac
import json
import time
from dataclasses import dataclass
from hashlib import sha256
from typing import Any, Literal

SIGNATURE_HEADER = "X-ReconSync-Signature"
EVENT_HEADER = "X-ReconSync-Event"
DELIVERY_HEADER = "X-ReconSync-Delivery"
DRILL_HEADER = "X-ReconSync-Drill"

#: How far a signature's timestamp may be from now.
DEFAULT_TOLERANCE_SECONDS = 300

Reason = Literal["malformed", "mismatch", "expired"]


class SignatureError(Exception):
    """The signature could not be trusted. Do not act on the payload."""

    def __init__(self, reason: Reason, message: str) -> None:
        super().__init__(message)
        self.reason: Reason = reason


def verify_signature(
    secret: str,
    header: str | None,
    body: bytes | str,
    *,
    now: float | None = None,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
) -> None:
    """Raise :class:`SignatureError` unless the signature is valid.

    ``body`` must be the **raw bytes as received**. Passing a re-serialised
    object is the most common way this fails in production: ``json.loads``
    followed by ``json.dumps`` changes separators and can reorder keys, and the
    signature covers the exact bytes that were sent. In Flask that means
    ``request.get_data()``, not ``request.get_json()``; in Django,
    ``request.body``.
    """
    timestamp, provided = _parse_header(header)

    if tolerance_seconds > 0:
        age = abs((time.time() if now is None else now) - timestamp)
        if age > tolerance_seconds:
            # Bounded replay: a request captured off the wire stops working
            # shortly after it was made.
            raise SignatureError(
                "expired",
                f"signature timestamp is {age:.0f}s away, outside the {tolerance_seconds}s tolerance",
            )

    raw = body.encode("utf-8") if isinstance(body, str) else body
    expected = hmac.new(
        secret.encode("utf-8"), f"{timestamp}.".encode("utf-8") + raw, sha256
    ).hexdigest()

    # compare_digest, not ==: a byte-by-byte comparison leaks how much of the
    # signature was correct, which is enough to forge one a byte at a time.
    if not hmac.compare_digest(expected, provided):
        raise SignatureError("mismatch", "signature does not match")


def _parse_header(header: str | None) -> tuple[int, str]:
    if not header:
        raise SignatureError("malformed", f"missing {SIGNATURE_HEADER} header")

    timestamp = 0
    signature = ""
    for part in header.split(","):
        key, sep, value = part.strip().partition("=")
        if not sep:
            raise SignatureError("malformed", f"cannot parse {SIGNATURE_HEADER}")
        if key == "t":
            try:
                timestamp = int(value)
            except ValueError as exc:
                raise SignatureError("malformed", "timestamp is not a number") from exc
        elif key == "v1":
            signature = value

    if not timestamp or not signature:
        raise SignatureError("malformed", f"{SIGNATURE_HEADER} is missing t or v1")
    return timestamp, signature


@dataclass(frozen=True)
class Webhook:
    """A verified webhook.

    The payload is kept as a plain dict as well as the fields worth naming,
    because events come in more than one shape and forcing every one through a
    single class would invent zero values for fields that do not apply.
    """

    event: str
    occurred_at: str
    data: dict[str, Any]
    raw: dict[str, Any]

    @property
    def transaction_id(self) -> str | None:
        return self.data.get("transaction_id")

    @property
    def advisory(self) -> bool:
        """Always true on a real event. Check your own ledger before paying."""
        return bool(self.data.get("advisory", False))

    @property
    def confidence(self) -> float:
        """0 to 1. Set your own bar rather than treating every verdict alike."""
        return float(self.data.get("confidence", 0.0))

    @property
    def is_drill(self) -> bool:
        """A fire drill. Acknowledge it and do nothing else."""
        return bool(self.data.get("drill", False))

    @property
    def amount_to_reverse_minor(self) -> int:
        """What is actually outstanding.

        Reverse this, never ``amount_minor``. When part of the money already
        reached the destination, refunding the full amount pays the customer
        twice for the part that arrived.
        """
        outstanding = self.data.get("outstanding_minor")
        if outstanding is not None:
            return int(outstanding)
        return int(self.data.get("amount_minor", 0))


def parse_webhook(
    secret: str,
    header: str | None,
    body: bytes | str,
    *,
    now: float | None = None,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
) -> Webhook:
    """Verify and parse, in the order that keeps you safe."""
    verify_signature(secret, header, body, now=now, tolerance_seconds=tolerance_seconds)

    payload = json.loads(body)
    return Webhook(
        event=payload.get("event", ""),
        occurred_at=payload.get("occurred_at", ""),
        data=payload.get("data", {}),
        raw=payload,
    )

File: sdk/node/src/webhook.ts
import { createHmac, timingSafeEqual } from "node:crypto";

// Verifying the signature is the one thing every receiver must get right.
//
// The payload advises reversing a customer's money. A handler that acts on an
// unverified body will act on anything anyone posts to that URL, and the whole
// design — advisory payloads, no credentials in the webhook — rests on the
// receiver checking who sent it.

export const SIGNATURE_HEADER = "x-reconsync-signature";
export const EVENT_HEADER = "x-reconsync-event";
export const DELIVERY_HEADER = "x-reconsync-delivery";
export const DRILL_HEADER = "x-reconsync-drill";

/** How far a signature's timestamp may be from now. */
export const DEFAULT_TOLERANCE_SECONDS = 300;

export class SignatureError extends Error {
  readonly reason: "malformed" | "mismatch" | "expired";

  constructor(reason: "malformed" | "mismatch" | "expired", message: string) {
    super(message);
    this.name = "SignatureError";
    this.reason = reason;
  }
}

/**
 * Verify a webhook signature.
 *
 * `body` must be the **raw bytes as received**. Passing a re-serialised object
 * is the most common way this fails in production: `JSON.parse` then
 * `JSON.stringify` can reorder keys and change whitespace, and the signature
 * covers the exact bytes that were sent. In Express that means
 * `express.raw({ type: "application/json" })` on this route, not `express.json()`.
 */
export function verifySignature(
  secret: string | string[],
  header: string | undefined | null,
  body: Buffer | string,
  options: { nowSeconds?: number; toleranceSeconds?: number } = {},
): void {
  const now = options.nowSeconds ?? Math.floor(Date.now() / 1000);
  const tolerance = options.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS;

  const { timestamp, signatures } = parseHeader(header);

  if (tolerance > 0 && Math.abs(now - timestamp) > tolerance) {
    // Bounded replay: a request captured off the wire stops working shortly
    // after it was made, so a recorded reversal cannot be replayed next week.
    throw new SignatureError(
      "expired",
      `signature timestamp is ${Math.abs(now - timestamp)}s away, outside the ${tolerance}s tolerance`,
    );
  }

  const raw = typeof body === "string" ? Buffer.from(body, "utf8") : body;
  // An array accepts a rotation from your side: hold the old and the new while
  // the sender changes over, then drop the old.
  const secrets = Array.isArray(secret) ? secret : [secret];

  let matched = false;
  for (const candidate of secrets) {
    const expected = Buffer.from(
      createHmac("sha256", candidate).update(`${timestamp}.`).update(raw).digest("hex"),
      "utf8",
    );

    for (const provided of signatures) {
      const given = Buffer.from(provided, "utf8");
      // Length first because timingSafeEqual throws on a mismatch; the
      // comparison itself is constant time, so a wrong signature does not
      // reveal how much of it was right. Every pair is checked even after one
      // matches, so the time taken does not depend on which one did.
      if (expected.length === given.length && timingSafeEqual(expected, given)) {
        matched = true;
      }
    }
  }

  if (!matched) {
    throw new SignatureError("mismatch", "signature does not match");
  }
}

/**
 * Reads the timestamp and **every** signature the header carries.
 *
 * Several v1 entries is the normal state while the sender rotates its secret,
 * so they are all collected. Keeping only the last would make the sender's
 * ordering decide whether your receiver worked.
 */
function parseHeader(header: string | undefined | null): {
  timestamp: number;
  signatures: string[];
} {
  if (!header) {
    throw new SignatureError("malformed", `missing ${SIGNATURE_HEADER} header`);
  }

  let timestamp = 0;
  const signatures: string[] = [];
  for (const part of header.split(",")) {
    const [key, value] = part.trim().split("=", 2);
    if (value === undefined) {
      throw new SignatureError("malformed", `cannot parse ${SIGNATURE_HEADER}`);
    }
    if (key === "t") {
      timestamp = Number.parseInt(value, 10);
      if (!Number.isFinite(timestamp)) {
        throw new SignatureError("malformed", "timestamp is not a number");
      }
    } else if (key === "v1" && value) {
      signatures.push(value);
    }
  }

  if (!timestamp || signatures.length === 0) {
    throw new SignatureError("malformed", `${SIGNATURE_HEADER} is missing t or v1`);
  }
  return { timestamp, signatures };
}

// --- payload shapes ---

export type EventType =
  | "reversal.triggered"
  | "reversal.completed"
  | "reversal.failed"
  | "transaction.suspect"
  | "transaction.reconciled"
  | "sla.at_risk"
  | "integration.silent"
  | "integration.recovered";

export interface EvidenceSignal {
  source: string;
  detail?: string;
  weight?: number;
}

export interface WebhookData {
  transaction_id: string;
  amount_minor: number;
  currency: string;
  reason?: string;

  /**
   * Present only when part of the money arrived. Reverse `outstanding_minor`,
   * never `amount_minor` — refunding the full amount when a fifth of it already
   * reached the destination pays the customer twice for that fifth.
   */
  credited_minor?: number;
  outstanding_minor?: number;

  debit_at: string;
  window_seconds: number;
  detected_at?: string;
  regulatory_deadline: string;

  /** Always true. Check your own ledger before moving money. */
  advisory: boolean;

  /** 0 to 1. Set your own bar rather than treating every verdict alike. */
  confidence: number;
  evidence?: EvidenceSignal[];

  seconds_until_breach?: number;

  /** Present only on a fire drill. Acknowledge it and do nothing else. */
  drill?: boolean;
}

export interface WebhookEnvelope {
  event: EventType;
  occurred_at: string;
  data: WebhookData;
}

/** Verify and parse in one step, in the order that keeps you safe. */
export function parseWebhook(
  secret: string | string[],
  header: string | undefined | null,
  body: Buffer | string,
  options?: { nowSeconds?: number; toleranceSeconds?: number },
): WebhookEnvelope {
  verifySignature(secret, header, body, options);
  return JSON.parse(typeof body === "string" ? body : body.toString("utf8")) as WebhookEnvelope;
}
